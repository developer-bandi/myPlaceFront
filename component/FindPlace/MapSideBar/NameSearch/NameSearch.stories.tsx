import { ComponentStory, ComponentMeta } from "@storybook/react";
import NameSearch from "./NameSearch";
import { action } from "@storybook/addon-actions";

export default {
  title: "FindPlace/MapSideBar/NameSearch",
  component: NameSearch,
  parameters: {
    layout: "fullscreen",
  },
} as ComponentMeta<typeof NameSearch>;

const Template: ComponentStory<typeof NameSearch> = (arg) => (
  <div style={{ width: 380, position: "relative" }}>
    <NameSearch {...arg} />
  </div>
);

export const generalType = Template.bind({});
generalType.args = {
  dispatchAddress: action("dispatchAddress"),
  dispatchSearchStore: action("dispatchSearchStore"),
};
