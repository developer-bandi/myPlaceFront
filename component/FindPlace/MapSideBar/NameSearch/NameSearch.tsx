import styles from "./NameSearch.module.scss";
import { AiOutlineSearch } from "react-icons/ai";
import { RefObject } from "react";

interface NameSearchProps {
  dispatchAddress: () => void;
  searchKeywordInputRef: RefObject<HTMLInputElement>;
  addressInputRef: RefObject<HTMLInputElement>;
  dispatchSearchStore: () => void;
}

const NameSearch = ({
  dispatchAddress,
  searchKeywordInputRef,
  addressInputRef,
  dispatchSearchStore,
}: NameSearchProps) => {
  return (
    <div className={styles.mainBlock}>
      <h1 className={styles.title}>기준 장소</h1>
      <div className={styles.inputBlock}>
        <input className={styles.adressInput} ref={addressInputRef} />
        <button
          className={styles.adressButton}
          onClick={() => {
            dispatchAddress();
          }}
        >
          설정
        </button>
      </div>
      <h1 className={styles.title}>검색 키워드</h1>
      <div className={styles.searchBlock}>
        <input className={styles.searchinput} ref={searchKeywordInputRef} />
        <button className={styles.searchbutton}>
          <AiOutlineSearch
            size={24}
            onClick={() => {
              dispatchSearchStore();
            }}
          />
        </button>
      </div>
    </div>
  );
};

export default NameSearch;
