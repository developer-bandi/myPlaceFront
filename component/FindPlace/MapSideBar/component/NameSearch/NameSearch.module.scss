.mainBlock {
  padding: 10px;
  border-bottom: 1px solid #e5e5e5;
  padding-bottom: 20px;
}

.searchBlock {
  width: 340px;
  border: 1px solid black;
  border-radius: 10px;
  display: flex;
  margin-left: 10px;
}

.searchinput {
  width: 295px;
  height: 40px;
  border-radius: 10px;
  font-size: 20px;
}

.searchbutton {
}

.inputBlock {
  padding-left: 10px;
  margin-bottom: 20px;
}

.adressInput {
  width: 290px;
  height: 30px;
  border: 1px solid black;
  border-radius: 5px;
  margin-right: 10px;
}

.adressButton {
  width: 50px;
  height: 30px;
  background-color: #00aaff;
  border-radius: 5px;
  color: white;
}

.title {
  padding-left: 10px;
  font-weight: bold;
  margin-bottom: 15px;
  margin-top: 10px;
  font-size: 16px;
}
