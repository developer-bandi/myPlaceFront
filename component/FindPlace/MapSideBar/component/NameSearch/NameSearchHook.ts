import {useEffect, useRef} from "react";
import {useDispatch, useSelector} from "react-redux";
import {RootReducer} from "../../../../../store";
import {setAdress} from "../../../../../store/reducers/hashtagSearchCondition/Reducer";
import {searchStore} from "../../../../../store/reducers/searchResult/Reducer";

const useNameSearch = () => {
  const dispatch = useDispatch();
  const searchKeywordInputRef = useRef<HTMLInputElement>(null);
  const addressInputRef = useRef<HTMLInputElement>(null);
  const settedAdress = useSelector(
    (state: RootReducer) => state.hashtagSearchCondition.adress
  );
  const settedKeyWord = useSelector(
    (state: RootReducer) => state.hashtagSearchCondition.keyword
  );

  useEffect(() => {
    if (addressInputRef.current !== null)
      addressInputRef.current.value = settedAdress.content;
  }, [settedAdress]);

  useEffect(() => {
    if (searchKeywordInputRef.current !== null) {
      searchKeywordInputRef.current.value = settedKeyWord;
    }
  }, []);

  const dispatchAddress = () => {
    if (addressInputRef.current !== null) {
      dispatch(setAdress({adress: addressInputRef.current.value}));
    }
  };

  const dispatchSearchStore = () => {
    if (searchKeywordInputRef.current !== null) {
      dispatch(
        searchStore({
          latitude: settedAdress.latitude,
          longitude: settedAdress.longitude,
          searchKeyword: searchKeywordInputRef.current.value,
        })
      );
    }
  };

  return {
    dispatchAddress,
    addressInputRef,
    searchKeywordInputRef,
    dispatchSearchStore,
  };
};

export default useNameSearch;
