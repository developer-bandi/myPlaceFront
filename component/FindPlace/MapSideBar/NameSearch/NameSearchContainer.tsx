import NameSearch from "./NameSearch";
import useNameSearch from "./NameSearchHook";

const NameSearchContainer = () => {
  const {
    dispatchAddress,
    addressInputRef,
    searchKeywordInputRef,
    dispatchSearchStore,
  } = useNameSearch();

  return (
    <NameSearch
      dispatchAddress={dispatchAddress}
      addressInputRef={addressInputRef}
      searchKeywordInputRef={searchKeywordInputRef}
      dispatchSearchStore={dispatchSearchStore}
    ></NameSearch>
  );
};

export default NameSearchContainer;
