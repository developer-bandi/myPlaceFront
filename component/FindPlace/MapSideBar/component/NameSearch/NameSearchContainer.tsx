import { useEffect, useRef } from "react";
import { useDispatch, useSelector } from "react-redux";
import { RootReducer } from "../../../../../store";
import { setAdress } from "../../../../../store/reducers/hashtagSearchCondition/hashtagSearchConditionReducer";
import { searchStore } from "../../../../../store/reducers/searchResult/searchResultReducer";
import NameSearch from "./NameSearch";

const NameSearchContainer = () => {
  const dispatch = useDispatch();
  const searchKeywordInputRef = useRef<HTMLInputElement>(null);
  const addressInputRef = useRef<HTMLInputElement>(null);
  const settedAdress = useSelector(
    (state: RootReducer) => state.hashtagSearchCondition.adress
  );
  const settedKeyWord = useSelector(
    (state: RootReducer) => state.hashtagSearchCondition.keyword
  );

  useEffect(() => {
    if (addressInputRef.current !== null)
      addressInputRef.current.value = settedAdress.content;
  }, [settedAdress]);

  useEffect(() => {
    if (searchKeywordInputRef.current !== null) {
      searchKeywordInputRef.current.value = settedKeyWord;
    }
  }, []);

  const dispatchAddress = () => {
    if (addressInputRef.current !== null) {
      dispatch(setAdress({ adress: addressInputRef.current.value }));
    }
  };

  const dispatchSearchStore = () => {
    if (searchKeywordInputRef.current !== null) {
      dispatch(
        searchStore({
          latitude: settedAdress.latitude,
          longitude: settedAdress.longitude,
          searchKeyword: searchKeywordInputRef.current.value,
        })
      );
    }
  };

  return (
    <NameSearch
      dispatchAddress={dispatchAddress}
      addressInputRef={addressInputRef}
      searchKeywordInputRef={searchKeywordInputRef}
      dispatchSearchStore={dispatchSearchStore}
    ></NameSearch>
  );
};

export default NameSearchContainer;
