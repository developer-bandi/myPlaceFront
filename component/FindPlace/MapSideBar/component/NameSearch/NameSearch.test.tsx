import {
  act,
  fireEvent,
  render,
  renderHook,
  screen,
} from "@testing-library/react";
import {Provider} from "react-redux";
import configureMockStore from "redux-mock-store";
import useNameSearch from "./NameSearchHook";
import newReact from "react";
import * as React from "react";
import NameSearch from "./NameSearch";

const mockStore = configureMockStore()({
  hashtagSearchCondition: {
    adress: {content: "testAdress"},
    keyword: "testKeyword",
  },
});
const wrapper = ({children}: any) => (
  <Provider store={mockStore}>{children}</Provider>
);

describe("NameSearch Hook 테스트", () => {
  it("처음 useEffect 테스트", () => {
    jest
      .spyOn(newReact, "useRef")
      .mockReturnValueOnce({current: {value: "test"}})
      .mockReturnValueOnce({current: {value: "test"}});
    const {result} = renderHook(() => useNameSearch(), {
      wrapper,
    });
    expect(result.current.addressInputRef).toEqual({
      current: {value: "testAdress"},
    });
    expect(result.current.searchKeywordInputRef).toEqual({
      current: {value: "testKeyword"},
    });
  });
  it("dispatchAddress함수 테스트", () => {
    jest
      .spyOn(newReact, "useRef")
      .mockReturnValueOnce({current: {value: "test"}})
      .mockReturnValueOnce({current: {value: "test"}});
    const {result} = renderHook(() => useNameSearch(), {
      wrapper,
    });
    act(() => {
      result.current.dispatchAddress();
    });
    expect(mockStore.getActions()[0]).toEqual({
      payload: {
        adress: "testAdress",
      },
      type: "hashtagSearchCondition/setAdress",
    });
  });
  it("dispatchSearchStore함수 테스트", () => {
    jest
      .spyOn(newReact, "useRef")
      .mockReturnValueOnce({current: {value: "test"}})
      .mockReturnValueOnce({current: {value: "test"}});
    const {result} = renderHook(() => useNameSearch(), {
      wrapper,
    });
    act(() => {
      result.current.dispatchAddress();
    });
    expect(mockStore.getActions()[1]).toEqual({
      payload: {
        adress: "testAdress",
      },
      type: "hashtagSearchCondition/setAdress",
    });
  });
});

describe("NameSearch Presentational 테스트", () => {
  it("정상 렌더링", () => {
    const dispatchAddressMock = jest.fn();
    const dispatchSearchStoreMock = jest.fn();
    const utils = render(
      <NameSearch
        dispatchAddress={dispatchAddressMock}
        searchKeywordInputRef={{current: null}}
        addressInputRef={{current: null}}
        dispatchSearchStore={dispatchSearchStoreMock}
      ></NameSearch>
    );
    expect(utils.container).toMatchSnapshot();
    fireEvent.click(screen.getByTestId("dispatchAddress"));
    fireEvent.click(screen.getByTestId("dispatchSearchStore"));
    expect(dispatchAddressMock).toBeCalledTimes(1);
    expect(dispatchSearchStoreMock).toBeCalledTimes(1);
  });
});
