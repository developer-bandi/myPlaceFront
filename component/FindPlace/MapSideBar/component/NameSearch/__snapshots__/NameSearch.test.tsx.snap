// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NameSearch Presentational 테스트 정상 렌더링 1`] = `
<div>
  <div
    class="mainBlock"
  >
    <h1
      class="title"
    >
      기준 장소
    </h1>
    <div
      class="inputBlock"
    >
      <input
        class="adressInput"
      />
      <button
        class="adressButton"
        data-testid="dispatchAddress"
      >
        설정
      </button>
    </div>
    <h1
      class="title"
    >
      검색 키워드
    </h1>
    <div
      class="searchBlock"
    >
      <input
        class="searchinput"
      />
      <button
        class="searchbutton"
      >
        <svg
          data-testid="dispatchSearchStore"
          fill="currentColor"
          height="24"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 1024 1024"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0 0 11.6 0l43.6-43.5a8.2 8.2 0 0 0 0-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
